### ISSUE TITLE
Feat: Request to call someone

### ISSUE THREAD

**Issue by gbonaspetti** [tracim/tracim#4843]
## Feature description and goals

- [ ] Add the possibility to call someone via his profile page with a button "Call".
- [ ] If a person is called, it displays a popup (Flash Message style) at the called person tab with the call request time and a link to a video conference room (pro.suricate.tv) specific to the two users. A sound is also played and a desktop notification showed.
- [ ] In the user's profile, you will also find a button "Your visio link with X" with a link to the room of this person.
- [ ] The link is automatically generated from the names of the users involved.
- [ ] The popup remains visible until it is closed.
- [ ] The link opens in a new tab.
- [ ] A bell appears in the respondent's browser tab.
- [ ] The time of the call should be the time the call was made with the clock of the person who should receive the call.
- [ ] Have a notification in the wall, something like "X has requested you for a live exchange".
- [ ] Add a flash message like "Call request sent. Go to _this link_ to access your video conference" when you call someone.
- [ ] Have options "Accept", "Decline", "On hold / In a moment".
- [ ] Have a timeout for the call of 5 min. The popup disappears and the sender is informed that his call has failed by a flash message.

## Prerequisites

None

## Extra information
A POC was made to test this feature. See the branch `poc/notification_improvements`.

**Comment by grignards** [comment]
Note: @lebouquetin told me we should support both jitsi and big blue button for this feature.
I had a quick look to big blue button, it will be more costly to add than jitsi as creating a room requires a login on the big blue button server (whereas for jitsi it's only a matter of generating an URL pointing to a meeting room).

**Comment by lebouquetin** [comment]
@grignards I expected BBB to be a bit more open ... I suggest to add BBB support in a future feature and to keep focus on jitsi meet integration for this issue.

**Comment by lebouquetin** [comment]
Eventually, if Galene can be used without login that would be an option - see https://linuxfr.org/news/galene-un-serveur-de-videoconference-libre

**Comment by lebouquetin** [comment]
The idea behind supporting BBB was to allow a choice (and so to implement something like a *generic* programming interface but it looks like it's not an easy case)

**Comment by grignards** [comment]
If I understand right, rooms are named "groups" in galene and can only be created by adding files on the server, see "Group definitions" in https://galene.org/README.html

**Comment by grignards** [comment]
I looked further, galene also has sub-groups which can be configured to be created automatically on access. Which means (if I understand right) that the administrator of a galene server would only have to create a root group (for example named tracim) and indicate this in Tracim's configuration.
But the documentation isn't totally explicit and the demo server doesn't allow sub-groups. So we should test it.

**Comment by charline-rageade** [comment]
@MariePrandstetter @grignards comment has been updated.

**Comment by charline-rageade** [comment]
@lebouquetin, we have two questions for you :
-  How do you manage the case where there are two people calling each other at the same time? 
-  When the called party accepts a call, is there an automatic redirection to the visio tab?

**Comment by lebouquetin** [comment]
> How do you manage the case where there are two people calling each other at the same time?

We don't with this first iteration. *but* the room must be the same (not like it is currently done). This may be done for example by building the room name like this:

1. sort usernames
2. build roomname="username1-username2"
3. build romname = md5sum(username1, username2)

We should ensure there are no side effects of specific usernames.

If A calls B at the same time than B calls A, they will fall into the same room which is ok (even if the workflow of accept/refuse is undefined)

**Comment by raphj** [comment]
md5 ou any hash would break the usage of the browser history to find the room and would make it hard to quickly join from another device. Wouldn't "username1-username2" be enough?

**Comment by grignards** [comment]
After discussing with @raphj  and @inkhey about the APIs, we settled on:

- POST `/users/<user_id>/outgoing_calls` (to create a new call request)
- PUT `users/<user_id>/outgoing_calls/<call_id>/state` (to change the state from the caller POV)
- PUT `users/<user_id>/incoming_calls/<call_id>/state` (to change the state from the callee POV)
- plus GET incoming/outgoing calls and individual calls.
The API could also be `/users/<user_id>/calls` mixing both incoming and outgoing calls. The semantics (and authorizations) would be slightly less explicit in this case, this is the main reason we chose to split it (and also by analogy with the followers/followed API which has more or less the same behavior).

**Comment by lebouquetin** [comment]
Explicit is better than implicit. I agree with the split :+1:

**Comment by lebouquetin** [comment]
The future call "a space room" would offer a similar API design? This is required in order to have something easy to discover?

**Comment by grignards** [comment]
> The future call "a space room" would offer a similar API design? This is required in order to have something easy to discover?

Yes, in fact I think the API could be the same for the callee/receiver/incoming calls. It is a bit different for the caller as the POST needs another parameter (the space's id) and that the `outgoing_calls` API should list only one call for a space room call.

**Comment by raphj** [comment]
@PhilippeAccorsi there is a change of default value for configuration web.notification.excluded (we add an exclusuib for `user_call.*`)

**Comment by raphj** [comment]
Tested on branch
